core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.status.api.v1
+
+import java.util.Date
+import javax.ws.rs.{GET, PathParam, Produces, QueryParam}
+import javax.ws.rs.core.MediaType
+
+import org.apache.spark.executor.{InputMetrics => InternalInputMetrics, OutputMetrics => InternalOutputMetrics, ShuffleReadMetrics => InternalShuffleReadMetrics, ShuffleWriteMetrics => InternalShuffleWriteMetrics, TaskMetrics => InternalTaskMetrics}
+import org.apache.spark.scheduler.{AccumulableInfo => InternalAccumulableInfo, StageInfo}
+import org.apache.spark.ui.SparkUI
+import org.apache.spark.ui.jobs.UIData.{StageUIData, TaskUIData}
+
+@Produces(Array(MediaType.APPLICATION_JSON))
+private[v1] class AllStagesResource(uiRoot: UIRoot) {
+
+  @GET
+  def stageList(
+    @PathParam("appId") appId: String,
+    @QueryParam("status") statuses: java.util.List[StageStatus]
+  ): Seq[StageData] = {
+    uiRoot.withSparkUI(appId) { ui =>
+      val listener = ui.stagesTab.listener
```

Comment:
Instead of accessing the `JobProgressListener` via the `stagesTab`, why not use `ui.jobProgressListener` instead?  If this will work everywhere, then I think we can avoid exposing the `stagesTab` like this.